### ISSUE TITLE
Inferreing key of mapped type and embedding into a string template type doesn't work in some cases

### ISSUE THREAD

**Issue by gauravgautamgoldcast** [microsoft/TypeScript#41196]
Writing this type in typescript 4.1:
```
type ExtractKeys<T extends { [k: string]: any }> = T extends infer G ? `x/${keyof G}` : never;
```
gives this error
```
main.ts:83:77 - error TS2322: Type 'keyof G' is not assignable to type 'string | number | bigint | boolean'.
  Type 'string | number | symbol' is not assignable to type 'string | number | bigint | boolean'.
    Type 'symbol' is not assignable to type 'string | number | bigint | boolean'.
      Type 'keyof G' is not assignable to type 'number'.
        Type 'string | number | symbol' is not assignable to type 'number'.
          Type 'string' is not assignable to type 'number'.

83 type ExtractKeys<T extends { [k: string]: any }> = T extends infer G ? `x/${keyof G}` : never;
```

But, splitting this type into two steps works:
```
type ExtractKeys<T extends { [k: string]: any }> = T extends infer G ? keyof G : never;
type Embedded = `x/${ExtractKeys<{s: 1, y: 2}>}`

const a: Embedded = "x/s";
```

**Comment by gauravgautamgoldcast** [comment]
I want to use this kind of inference:
```
type GetterKeys<T extends Record<string, { getters: { [k: string]: any} }>> =
  T extends Record<infer K, { getters: infer G }> ? `${K}/${keyof G}` : never;
```

```
const modtree = {
   mod1: {
      getters: {
         val1(state) {
            return state.x;
         },
         val2(state) {
            return state.y;
         }
      }
   },
   mod2: {
      getters: {
         val1(state) {
           return state.x;
         }
       }
    }
}
```

then 
```
GetterKeys<typeof modtree> -> 'mod1/val1' | 'mod1/val2' | 'mod2/val1';
```

**Comment by RyanCavanaugh** [comment]
This is a correct error; what you want to write is
```ts
type ExtractKeys<T extends { [k: string]: any }> = T extends infer G ? `x/${(string & keyof G)}` : never;
```
if you want to discard symbol keys.

**Comment by johncmunson** [comment]
Could someone please explain what's going on here?

I arrived at this Github issue because I was having trouble getting the `keyof` operator to work with generics when used inside of a template literal type.

What I thought would work...
```ts
type SortBy<T extends {[key: string]: any}> = `${keyof T}.${'asc' | 'desc'}`
```
What actually worked, after reading @RyanCavanaugh's suggestion above...
```ts
type SortBy<T extends {[key: string]: any}> = T extends infer G ? `${(string & keyof G)}.${'asc' | 'desc'}` : never
```

The crux of the issue seems to be that `keyof` works just fine inside of template literal types, but only if you aren't using generics.

Without generics...
```ts
// Both types compile and they are equivalent
type Foo = keyof {foo: 'bar'}
type Foo = `${keyof {foo: 'bar'}}`
```
With generics...
```ts
type Foo<T> = keyof T       // Compiles
type Foo<T> = `${keyof T}`  // ERROR
```

This seems to me like a broken abstraction and that concepts like `keyof`, generics, and template literal types are not staying completely orthogonal of each other. I'm happy that I was able to get my use case working, but the code that I needed to resort to seems very obtuse.

**Comment by qwen-3108** [comment]
@johncmunson Thank you! Your response highlighted what I missed from @RyanCavanaugh's and saved me hours!

For others who come across this thread, I am by no means a typescript expert, but after some probing, I think I might be able to provide some explanation on why template literals behave like this with generics.

I think it is related to the evaluation behavior of the `extends` keyword.

**TL;DR**: The `extends` keyword does not actually narrow down the generic's keys to only string keys.

---

**Detailed response**

Consider the below
```
type J = {[key: string]:any};
type K = {[key: symbol]: any};
type L = {[key: symbol | string]: any};
type H = {foo: 'bar'}; // concrete type
```
Apparently, `K extends J` and `L extends J` both evaluates to `true` in the TS playground, even though they contain `symbol` keys and type `J` explicitly states that it only has `string` keys.

However, `K extends H` and `L extends H` are able to properly evaluate to `false`.
 
[TS playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAUlC8UDeBtA1hEAuAzsATgJYB2A5gLpYCGxIAvgNwBQokUA0gsupllDiAC2AIwD2AG0pQa9Zq2gAZBKgzZ+QseKgAffgRIU+Mxi3DQAElyQAzUaL4ByYVXwPGUAPQeoAY1HEffAhgaHkmcPkoELwARi5OCAAPEOIAExxYKAB+KAAiAgBXCFyoPlzrKnEcYrkzKIg8ACYuJSSU9Myc-PwikrKKqpqoCLro4ABmeKg2iDSMyy7C4tK8gerc2rYxgBYW6eTZjoW8pb7VyvXmKCggA)
```
type test1 = K extends J ? "true" : "false";  // true
type test2 = L extends J ? "true" : "false";  // true

type test3 = K extends H ? "true" : "false"; // false
type test4 = L extends H ? "true" : "false"; // false
```
Thus,
```
type Foo = `${keyof {foo: 'bar'}}`
```
^ This compiles because the compiler is able to conclusively deduce that there'd be no symbol keys.

```
type Foo<T extends {[key: string]:any}> = `${keyof T}` // ERROR
```
^ But this errors because the `extends` keyword does not actually narrow down the type of T's keys to only string.

And this is why we need to manually apply the `string &` constraint to tell the compiler we'll leave out symbol keys (template literals does not take `symbol` type as argument).
```
type Foo<T extends {[key: string]:any}> = `${string & keyof T}` // compiles
```
